### ISSUE TITLE
Random() function returning a value between 0 and 128-1 but "max" parameter is a char

### ISSUE THREAD

**Issue by lunoka** [pce-devel/huc#35]
The random function is limited to return a char between 0 & 128-1, is there a technical reason for this clamp whereas it takes a char parameter? this limitation was not specified in the doc ( from ArchaicPixels which is now down ... ) and very misleading.

**Comment by jbrandwood** [comment]
Is a char signed in HuC?

Anyway, this function has not changed since HuC v1.5 ...

; ----
; random
; ----
; return a random number in the interval 0 <= x < A
; ----
; IN :	A = range (1 - 128)
; ----
; OUT : A = random number
; ----
;

**Comment by jbrandwood** [comment]
Oh, and it's an absolutely _terrible_ random number function anyway, you'd be much better off writing your own.

**Comment by lunoka** [comment]
Ha, alright, it will do the trick anyway, I will check that on next project then, thank you.

**Comment by lunoka** [comment]
I think I'll need some update because I have absolutely no idea what HuCC is ... ^^; are you rewriting the all core and compiler?

**Comment by jbrandwood** [comment]
Basically "yes"! :-)

It's not been announced yet, because there are more impovements that I still to make, and because it needs both examples and documentation ... but some current HuC developers are already using it to replace the old compiler, and I just checked-in the  compiler source and include directory into this repo.

If you're not using a lot of inline asm, then the transition should be pretty simple. The only issue that you'd likely find is if the new libraries don't support one of the HuC functions that you're using. If that's the case, just let me know, and I can see about adding it (or explaining how to get around the need for it).

If you are using a lot of inline asm, or writing you own asm procedures, then it may take a little more work to use HuCC, but the changes are easy to make, and still have the code work with both HuC and HuCC.

**Comment by jbrandwood** [comment]
The most difficult thing that a project needed to change so far was Space Ava's use of "dx" as a C variable name, which is no longer allowed because HuCC reverts to the old HuC v3.21 behavior of having "_dx" be reserved as an asm variable for System Card routines.

**Comment by lunoka** [comment]
Awesome!

I could try to build my puzzle game with this upgrade, is it just a matter of editing the Make_ex.inc to point to HuCC binary?
CC     = $(BINDIR)/hucc

**Comment by jbrandwood** [comment]
You'll also need to change PCE_INCLUDE to use "include/hucc" instead of "include/huc", and there may be some command line options that need to change, especially if your build script calls PCEAS seperately, in which case you must use the new "--hucc" option to PCEAS.

**Comment by jbrandwood** [comment]
Oh, one other thing, HuCC only supports a small stack, so if you're not already using "-msmall" with HuC, your project is unlikely to work.

**Comment by lunoka** [comment]
![error](https://github.com/user-attachments/assets/192131c3-da68-4033-97ea-a235e6c93d38)
Ok, got this on pceas side.

**Comment by jbrandwood** [comment]
Ah, those look like they're from inline asm. You're using some of the old MagiKit asm macros that I've not included.

I can add them quite quickly, or you can define them in a header file yourself (I will add them but it'll take about 30 minutes).

The other thing is that the long-branch macros like "lbne" won't be supported. In HuCC you just use a regular "bne" and PCEAS will automatically turn it into a long-branch if it needs to.

To keep the source compatible with both compilers, you just put the lbne instruction in a ".ifdef HUCC" conditional section.

**Comment by jbrandwood** [comment]
Although ... if you're using that lbne to branch outside of the current #asm section but within the same C function, then you may find that it doesn't work. That would be a very ugly thing to do in C code.

**Comment by lunoka** [comment]
Ha, yes. Touko made me an asm fadeout function with inline asm ( and also a clear vram function ). I forgot about that.

**Comment by jbrandwood** [comment]
If it's branching within the same asm section, it should work fine. If it's doing something "strange", then you can send me the function and I'll rewrite it in a way that will work with both compilers.

**Comment by jbrandwood** [comment]
And yep, HuCC really needs to include a new library function for palette fading, it's such a common requirement in games.

**Comment by lunoka** [comment]
alright, I'll wait for you to add the missing macros. If I get stuck, I'll send you the fadeout file in pm on proboard forum, thank you :)

**Comment by jbrandwood** [comment]
I've checked-in the additional macros, including the "lbne" (I guess it's easier that way).

It'll take github about 20 minutes to make a new build for you to download.

**Comment by lunoka** [comment]
Going further now, it seems to have a missing function : 
![image](https://github.com/user-attachments/assets/e7756f29-e26f-4868-8ab5-cd1ac78427b2)

**Comment by jbrandwood** [comment]
OK, thanks for your patience! I'll add the missing function in the next few hours and let you know when a new build is available.

Best wishes,

John

**Comment by lunoka** [comment]
no problem, take your time, my game is already complete and the rom is available.
Btw, have you an idea about possible speed or size improvement of HuCC towards HuC?

**Comment by jbrandwood** [comment]
I've added the function, and github has a new build ready that you can try. :-)

I think that the speed and size improvement is going to vary quite a lot from project to project, depending upon how much use they make of the different C features that have been improved, sometimes a lot, sometimes very little.

Unused library functions are automatically stripped out, which also helps, on top of the complete removal of library functions that developers just don't seem to have ever used in the projects that I've tested (such as the put_tile() where you're the first project to use it ... so now it is included).

You'll have to tell me just how much memory is saved, and whether the improved code-generation is actually something that you find noticeable. At the end of the day, if your game already runs at 60Hz, then it'll be difficult to actually see any speedup unless you're using timing-bars or some other metric to actually measure the code speed.

Best wishes,

John

**Comment by lunoka** [comment]
thank you! Now the game is fully building but the sound part is messed up and the game is crashing after some time.
I may add that I have some custom code TBXray made for me to manage far pointers for my songs ( I made my own custom sound engine from scratch ), do you think it could come from this since I don't think you add the chance to see this code.

**Comment by lunoka** [comment]
I sent to you in pm on proboard forum the 2 files made by TBXRay.

**Comment by jbrandwood** [comment]
Yes, it could be those functions, I would need to see them. HuCC has a slightly different calling method for asm functions than HuC, so that's definitely where problems can occur.

HuCC is also a bit stricter about requiring function declarations than HuC.

Your message hasn't come through yet on the forum, did you send it to "elmer"?

**Comment by lunoka** [comment]
woopsie, it should be ok now,sorry.
